{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac2727ef-8e7e-4e19-997e-371b4ca5f6b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO\n",
    "# Logic for pushing the playbook generated to the git repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b128be88-0353-4718-a0d5-70cba7317c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import shutil\n",
    "import os\n",
    "\n",
    "GITHUB_API_KEY = os.getenv(\"github_api_key\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76d30f50-be0c-4d22-8fb0-ef9e6fbc465b",
   "metadata": {},
   "outputs": [],
   "source": [
    "LOCAL_REPO_PATH = \"AIOps-for-Error-Remediation/pipeline\" # For pipeline runtime use\n",
    "PLAYBOOK_PATH = \"/opt/app-root/src/AIOps-for-Error-Remediation/pipeline\"\n",
    "GIT_USERNAME = \"wj-tan\"\n",
    "GIT_EMAIL = \"tan.weijian1996@gmail.com\"\n",
    "COMMIT_MESSAGE = \"Remediation playbook commit\"\n",
    "BRANCH_NAME = \"main\"  # or 'master' depending on your repo\n",
    "REPO_URL = f\"https://{GIT_USERNAME}:{GITHUB_API_KEY}@github.com/beedydo/AIOps-for-Error-Remediation.git\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cd9e54c2-de3b-4f0a-9b01-f204cd044f5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['git', 'config', '--global', 'user.email', 'tan.weijian1996@gmail.com'], returncode=0)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure Git user config is set (optional if already globally set)\n",
    "subprocess.run([\"git\", \"config\", \"--global\", \"user.name\", GIT_USERNAME])\n",
    "subprocess.run([\"git\", \"config\", \"--global\", \"user.email\", GIT_EMAIL])\n",
    "\n",
    "# Clone the AIOps Repo (Needed for the pipeline runtime)\n",
    "subprocess.run([\"git\", \"clone\", REPO_URL], check=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e002da73-16a4-4444-a46b-8dc4dcc628a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['git', 'remote', 'set-url', 'origin', 'https://wj-tan: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@github.com/beedydo/AIOps-for-Error-Remediation.git'], returncode=0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set remote URL with credentials\n",
    "#subprocess.run([\"git\", \"remote\", \"set-url\", \"origin\", REPO_URL], cwd=PLAYBOOK_PATH)\n",
    "subprocess.run([\"git\", \"remote\", \"set-url\", \"origin\", REPO_URL], cwd=LOCAL_REPO_PATH) # For pipeline runtime use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "169fb56b-2d85-4ed8-bdc3-808a4459e307",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'AIOps-for-Error-Remediation/pipeline/remediation.yaml'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[11]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Copy remediation.yaml from previous process into the repo pipeline folder\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43mshutil\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mremediation.yaml\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m.\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mLOCAL_REPO_PATH\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mremediation.yaml\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/usr/lib64/python3.11/shutil.py:431\u001b[39m, in \u001b[36mcopy\u001b[39m\u001b[34m(src, dst, follow_symlinks)\u001b[39m\n\u001b[32m    429\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m os.path.isdir(dst):\n\u001b[32m    430\u001b[39m     dst = os.path.join(dst, os.path.basename(src))\n\u001b[32m--> \u001b[39m\u001b[32m431\u001b[39m \u001b[43mcopyfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdst\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfollow_symlinks\u001b[49m\u001b[43m=\u001b[49m\u001b[43mfollow_symlinks\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    432\u001b[39m copymode(src, dst, follow_symlinks=follow_symlinks)\n\u001b[32m    433\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m dst\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/usr/lib64/python3.11/shutil.py:258\u001b[39m, in \u001b[36mcopyfile\u001b[39m\u001b[34m(src, dst, follow_symlinks)\u001b[39m\n\u001b[32m    256\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(src, \u001b[33m'\u001b[39m\u001b[33mrb\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fsrc:\n\u001b[32m    257\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m258\u001b[39m         \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mdst\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mwb\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m fdst:\n\u001b[32m    259\u001b[39m             \u001b[38;5;66;03m# macOS\u001b[39;00m\n\u001b[32m    260\u001b[39m             \u001b[38;5;28;01mif\u001b[39;00m _HAS_FCOPYFILE:\n\u001b[32m    261\u001b[39m                 \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: 'AIOps-for-Error-Remediation/pipeline/remediation.yaml'"
     ]
    }
   ],
   "source": [
    "# Copy remediation.yaml from previous process into the repo pipeline folder\n",
    "#shutil.copy(\"remediation.yaml\", os.path.join(LOCAL_REPO_PATH, \"remediation.yaml\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "14a41a62-9b1c-4011-bd1e-4e60fe71af09",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "fatal: The permissions on your socket directory are too loose; other\n",
      "users may be able to read your cached credentials. Consider running:\n",
      "\n",
      "\tchmod 0700 /opt/app-root/src/.git-credential-cache\n",
      "fatal: cache daemon did not start: \n",
      "From https://github.com/beedydo/AIOps-for-Error-Remediation\n",
      " * branch            main       -> FETCH_HEAD\n",
      "   1a4cc90..85c3908  main       -> origin/main\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updating 1a4cc90..85c3908\n",
      "Fast-forward\n",
      " rulebooks/inventory                |  3 +++\n",
      " rulebooks/webserver_rulebook.yml   | 42 ++++++++++++++++++++++++++++----------\n",
      " rulebooks/website_index_down.yml   | 17 +++++++++++++++\n",
      " rulebooks/website_service_down.yml |  0\n",
      " 4 files changed, 51 insertions(+), 11 deletions(-)\n",
      " create mode 100644 rulebooks/inventory\n",
      " create mode 100644 rulebooks/website_index_down.yml\n",
      " create mode 100644 rulebooks/website_service_down.yml\n",
      "[main b7c235f] Remediation playbook commit\n",
      " 1 file changed, 36 insertions(+), 15 deletions(-)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "fatal: The permissions on your socket directory are too loose; other\n",
      "users may be able to read your cached credentials. Consider running:\n",
      "\n",
      "\tchmod 0700 /opt/app-root/src/.git-credential-cache\n",
      "fatal: cache daemon did not start: \n",
      "To https://github.com/beedydo/AIOps-for-Error-Remediation.git\n",
      "   85c3908..b7c235f  main -> main\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['git', 'push', 'origin', 'main'], returncode=0)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run Git commands to pull, add, commit, and push\n",
    "subprocess.run([\"git\", \"pull\", \"origin\", BRANCH_NAME], cwd=PLAYBOOK_PATH)\n",
    "subprocess.run([\"git\", \"add\", \"remediation.yaml\"], cwd=PLAYBOOK_PATH)\n",
    "subprocess.run([\"git\", \"commit\", \"-m\", COMMIT_MESSAGE], cwd=PLAYBOOK_PATH)\n",
    "subprocess.run([\"git\", \"push\", \"origin\", BRANCH_NAME], cwd=PLAYBOOK_PATH)\n",
    "\n",
    "# subprocess.run([\"git\", \"pull\", \"origin\", BRANCH_NAME], cwd=LOCAL_REPO_PATH)\n",
    "# subprocess.run([\"git\", \"add\", \"remediation.yaml\"], cwd=LOCAL_REPO_PATH)\n",
    "# subprocess.run([\"git\", \"commit\", \"-m\", COMMIT_MESSAGE], cwd=LOCAL_REPO_PATH)\n",
    "# subprocess.run([\"git\", \"push\", \"origin\", BRANCH_NAME], cwd=LOCAL_REPO_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d09928b-9056-4d98-83eb-4b121cc4694a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
